### ISSUE TITLE
"evil-search-forward" freezes emacs sometime in org-mode

### ISSUE THREAD

**Issue by StreakyCobra** [syl20bnr/spacemacs#3623]
To reproduce:
1. Open the `*scratch*` buffer: `SPC b s`
2. Paste this line in it:
   
   ```
   [[path/is/problematic][link]] is problematic
   ```
3. Change the buffer to `org-mode`: `SPC : org-mode RET`
4. Search for `is/` or `path`

Notice:
- Doing the same but staying in `fundamental-mode` doesn't have this behaviour.
- It appears when all the matches are only inside links' path.
- If we search with `evil-search-forward` in `fundamental-mode`, then switch to `org-mode`, doing `evil-search-next` doesn't freeze.

So for now the rule seems to be: «Using `evil-search-forward` or `evil-search-backward` in org-mode for an expression which is only inside links' path, freeze emacs.».

It's probably an upstream bug, but as I'm not used to standard emacs enough to test it without spacemacs, I can't exclude it's a spacemacs bug. Maybe someone can try?

Bug reproduced on both `master` and `develop` fresh installs (vim mode at least), emacs 24.5.1.

**Comment by manfredlotz** [comment]
Some weeks ago I had opened issue #3338  which reported a freeze when searching for the word 'link' in the README.org buffer. Funny thing was that the freeze happened when opening the file via 'SPC f e h'. If later I repoen the file via 'SPC f e h' there is no freeze.

Don't know if this is the same thing.

**Comment by manfredlotz** [comment]
In issue #3338 I added an append because I found that emacs antialiasing might be the problem of the freezes.

**Comment by manfredlotz** [comment]
Would be interesting which font is defined in .spacemacs of StreakyCobra. In my situation (#3338) changing the font, and no freeze in the described situation (SPC feh and searching for link in README.org)

**Comment by manfredlotz** [comment]
Clarifying: As StreakyCobra said the problem shows up with 'emacs -nw' it has nothing to do with issue #3338.

**Comment by Alexander-Shukaev** [comment]
Profiler trace shows the following:

```
                   - isearch-search                             23604  91%
                    - isearch-filter-visible                    14667  56%
                       isearch-range-invisible                  12937  50%
                    + isearch-search-string                      8009  30%
```

If only I knew where to start debugging this...

**Comment by Alexander-Shukaev** [comment]
I think I've found a solution.  Just want to note one thing.  I'm not a `spacemacs` user and I also have completely reimplemented the `evil-search` module based on `isearch` as the original contained numerous subtle bugs.  In fact, I have a large custom layer on top of `evil` which hooks certain parts of it, extends, replaces, or removes them completely.  In future, I plan to fork and deprecate Evil entirely and release a new package to the public when I have more spare time (the idea is actually to produce a more generic and bug-free version of Evil).  Anyway, since some of the code of this rewritten `evil-search` module is still in common, I also have these freezing issues.  They were annoying me for quite some time now, so I decided to nail it.  It looks like if you say

```
(setq-default search-invisible t)
```

the problem should be gone, at least it does for me.  Give it a shot.
